Add `Expr(:ivdepscope)` to support not marking the entire loop body as `ivdep`

N5N3 opened the pull request:
- **Motivation:**
Currently, `@simd ivdep` assumes the entire loop is free of "no loop-carried memory dependencies", which limits its usage in our broadcast system.
This PR tries to split `julia.ivdep` into 2 meta: `julia.ivdep.begin` and `julia.ivdep.end`, and makes the simd-loop pass only marks the access within a `begin/end` block as `MD_mem_parallel_loop_access`.
With this PR, if we find that all the args in a flat `bc::Broadcasted` are safe to parallelly loaded, and the `dest::AbstractArray` is safe to parallelly strored.
Then we can implement the `copyto!` kernal as:
```julia
@eval function copyto!(dest::AbstractArray, bc::Broadcasted)
    @inbounds @simd for I in eachindex(bc)
        $(Expr(:loopinfo, Symbol("julia.ivdep.begin")))    # args are safe to load
        args = _getindex(bc.args, I)
        $(Expr(:loopinfo, Symbol("julia.ivdep.end")))
        temp = bc.f(args...)                                   # bc.f might have side effect
        $(Expr(:loopinfo, Symbol("julia.ivdep.begin")))    # result is safe to store
        dest[I] = temp
        $(Expr(:loopinfo, Symbol("julia.ivdep.end")))
    end
end
```
If `bc.f` is free of memory access, then LLVM should FIND this loop vectorlizable and add no runtime check. (and we can make `a .+= 1` vectorlized more easily)
If not, then let LLVM checks whether `bc.f` might have side effect.

- **Changes in this PR:**
This PR only changes the pass inplementation.
And makes `@simd ivdep` generates a `julia.ivdep.begin/end` block instead of a single `julia.ivdep`
The usage and effect of `@simd` and `@simd ivdep` are not changed.

I'm not familiar with LLVM and I'm not sure this change is the correct way to make self-inplace broadcast vectorlizable.
All suggestions and comments are welcome.

N5N3 commented:
After some further trial with:
1. 06a16b84e08b59897e06e9c1d80df710c5938b8c : removes `@inline`, and replace `Ref` with `Some`, thanks @vtjnash's code
2. 7320cc92f0c59b269de1e42540eac22e53c4c09a : implements the broadcast kernal with local ivdep marks and some optimization inspired by #30973.

I find that, to make self-inplace broadcast vectorlizable, we only need to tell LLVM "`arrayref` within `bc`'s `getindex` and `arrayset`  within `dest`'s `setindex!` are parallelly-safe during the broadcast kernal". Combining this PR with the above 2 commits, on my desktop I have:
```julia
julia> using BenchmarkTools
julia> a = zeros(Float32, 128, 32); a_ = view(a, 1:128, 1:32);
julia> @btime $a .+= $a;
  280.000 ns (0 allocations: 0 bytes)  # on 1.7.0: 1.650 μs (0 allocations: 0 bytes)
julia> @btime $a .+= $a .+ $a .+ $a;
  511.979 ns (0 allocations: 0 bytes)  # on 1.7.0: 3.550 μs (0 allocations: 0 bytes) 
julia> @btime $a_ .+= $a_;
  289.286 ns (0 allocations: 0 bytes)  # on 1.7.0: 1.730 μs (0 allocations: 0 bytes)
julia> @btime $a_ .+= $a_ .+ $a_ .+ $a_;
  609.551 ns (0 allocations: 0 bytes)  # on 1.7.0: 6.440 μs (0 allocations: 0 bytes)
```
Some simple safety check:
```julia
julia> const p = Ref(0);
julia> a = zeros(Float32, 128, 32); b = similar(a);

julia> f(x) = x + (p[] += 0); # f has no side-effect
julia> @btime $b .= f.($a);
  221.526 ns (0 allocations: 0 bytes) # on 1.7.0: 1.440 μs (0 allocations: 0 bytes) 
julia> @btime $a .= f.($a);
  164.899 ns (0 allocations: 0 bytes) # on 1.7.0: 2.033 μs (0 allocations: 0 bytes)

julia> g(x) = x + (p[] = ~p[]); # g has side-effect
julia> @btime $b .= g.($a);
  2.411 μs (0 allocations: 0 bytes) # on 1.7.0: 2.522 μs (0 allocations: 0 bytes)
julia> @btime $a .= g.($a);
  2.411 μs (0 allocations: 0 bytes) # on 1.7.0: 2.511 μs (0 allocations: 0 bytes)
```
The above example shows that this change is safer than replacing `@simd` in broadcast kernal with `@simd ivdep`.
But I didn't look into why 1.7.0 faild to vectorlize `b .= f.(a)`. Maybe more tests are needed.
Hope this is the correct way to "fix" #43153 safely.

vchuravy commented on base/simdloop.jl:
```diff
@@ -125,12 +126,12 @@ either case, your inner loop should have the following properties to allow vecto
 """
 macro simd(forloop)
-    esc(compile(forloop, nothing))
+    esc(compile(forloop, Symbol("julia.ivdep.end")))
```
Can we keep this `nothing`? Makes little sense to have non-matching "begin"/"end" construct

N5N3 replied:
Didn't know that we have `Base.Experimental.Const` and `@aliasscope`, which seems a better tool to make `a .+= 1` vectorlizable.
So the very first motivation for this PR no longer holds: (If user wants partial `ivdep`, they could use `@aliasscope` instead in most cases.)
Then the second question is, can we use `Experimental` api in our `Base`?

N5N3 replied:
IIUC, we'd better have something like `Expr(:ivdepscope)` and `Expr(:popivdepscope)` instead.
(and add a `@ivdep` macro ?).

vchuravy reviewed (commented):
I am not a fan of the current construct. Currently `Expr(:loopinfo)` has the semantics that it should always be the last instruction in a loop. For a `begin/end` construct I would rather mimic `Base.Experimental.Const` and `@aliasscope`

vchuravy reviewed (changes requested):
`Expr(:loopinfo)` has the dedicated meaning to attach metadata to the LLVM loop. The current `ivdep` implementation uses that since it is for the entire scope of the loop. See #31376 for another use-case.

For scoped ivdep I think the implementation should mirror more #31018

N5N3 commented:
IIUC，we dont want another Const like wrapper.
I know little about LLVM, but it looks like the parallel meta must be assigned with loop Id.
we still need to fallback to the current loop marker to get that.

Replacing “loopinfo, Julia.ivdep” with something like “ivdepscope, begin/end” might make sense.
I tried by adding a renamed “loopinfo” but failed. (See above reply). In fact, I never touched Julia’s
underlying part before, so I must miss something.

N5N3 committed:
add `Expr(:ivdepscope)`

1. introduce `jl_ivdepscope_sym`
2. define `jl_ivdepscope_error` to thrown error message.

N5N3 committed:
modify `codegen`

make `Expr(:ivdepscope, :begin/end)` lowered to `jl_ivdepscope_func`

N5N3 committed:
modify pass

1. let `loopinfo_mark` erase `julia.ivdepscope` if it has `julia.simd`.
2. erase `julia.ivdepscope` in unreachable branch even there's no `loopinfo_mark`. (make error message clearer)

N5N3 committed:
add `@ivdep`

N5N3 commented:
I'm not sure whether this is the correct way to implement scoped `ivdep`:
1. A new `meta` head `Expr(:ivdepscope)` is introduced, which will be lowered as `call void @julia.ivdepscope()` with `begin/end` Metadata:
2. When `loopinfo_marker` found `julia.simdloop`, it will add `parallelaccess` meta based on a simple `begin/end` count.
During which the `call void @julia.ivdepscope()` will also be erased. 
3. If there're remaining `julia.ivdepscope()` after all optimizaion, user might use `@ivdep` outside `@simd for end`. So I make `@julia.ivdepscope()` fallback to `jl_ivdepscope_error` to throw error.
4. User might write invalid code and thus leads to unreachable branch. To recover the original error message, the simdpass will erase `call void @julia.ivdepscope()` within basicblocks ending with `unreachable `

N5N3 committed:
add throw test

N5N3 commented:
I guess we won't need this after #43852.
